hw03_frequency_analysis/top.go

```diff
@@ -1,6 +1,44 @@
 package hw03frequencyanalysis
 
-func Top10(_ string) []string {
-	// Place your code here.
-	return nil
+import (
+	"math"
+	"sort"
+	"strings"
+)
+
+func Top10(s string) []string {
+	wordsCount := map[string]int{}
+	words := make([]string, 0, len(wordsCount))
+
+	// Заполняем словарь (слово - кол-во в тексте)
+	for _, v := range strings.Fields(s) {
+		key := prepareKey(v)
+		wordsCount[key]++
+	}
+
+	// Преобразуем словарь в слайс со словами
+	for word := range wordsCount {
+		if word == "-" {
+			continue
+		}
+		words = append(words, word)
+	}
+
+	// Сортируем слайс по кол-ву слов в тексте
+	sort.Slice(words, func(i int, j int) bool {
+		// Если кол-во слов равно, то сравниваем сами слова (лексикографическая сортировка)
+		if wordsCount[words[i]] == wordsCount[words[j]] {
+			return words[i] < words[j]
+		}
+		return wordsCount[words[i]] > wordsCount[words[j]]
+	})
+
+	// На случай, если слов в тексте меньше 10, высчитываем длину итогового слайса
+	sliceLen := int(math.Min(float64(len(words)), 10))
```

Comment:
не очень ясна необходимость использовать такую конструкцию. len же уже возвращает целочисленное значение

Reply:
math.Min() принимает на вход и возвращает значение с типом float64

Reply:
Можно было сделать через if и избежать лишних конвертаций, но так мне показалось лаконичнее.

Reply:
Это я понимаю. Я не понимаю чем ваше выражение лучше `sliceLen := len(words)`

Reply:
По условию задачи необходимо выводить top 10 слов. Это выражение на случай, когда слов в тексте меньше 10.
Как вариант :
```suggestion
        sliceLen := len(words)
        if sliceLen  > 10 {
	    sliceLen =  10
	}
```